File: 2022/aoc/day1.solver.ts
import { check, getProblemInput, parseBlocks, parseNumbers } from './utils'

const day = 1

const sample = `
    1000
    2000
    3000

    4000

    5000
    6000

    7000
    8000
    9000

    10000`

const input = getProblemInput(day)

const solve = (text, count) => parseBlocks(text)
    .map(block => parseNumbers(block).sum())
    .sort((a, b) => b - a)
    .take(count)
    .sum()

const solve1 = text => solve(text, 1)
const solve2 = text => solve(text, 3)

check(`Day ${day}.1 Sample`,  () => solve1(sample),  24000)
check(`Day ${day}.1 Problem`, () => solve1(input),   69177)
check(`Day ${day}.2 Sample`,  () => solve2(sample),  45000)
check(`Day ${day}.2 Problem`, () => solve2(input),  207456)

File: 2023/aoc/day1.solver.jl
day = 1
include("utils.jl")

input = getProblemInput()

sample1 = [
    "1abc2",
    "pqr3stu8vwx",
    "a1b2c3d4e5f",
    "treb7uchet",
]

solve1 = lines ->
    @pipe lines |>
    map(l -> replace(l, r"\D+" => ""), _) |>
    map(l -> parse(Int, l[begin]*l[end]), _) |>
    sum

check("Day $day.1 Sample",  () -> solve1(sample1), 142)
check("Day $day.1 Problem", () -> solve1(input), 54953)

sample2 = [
    "two1nine",
    "eightwothree",
    "abcone2threexyz",
    "xtwone3four",
    "4nineeightseven2",
    "zoneight234",
    "7pqrstsixteen",
]

function solve2(lines)
    words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    function fixup(line)
        c = 1
        while c <= length(line)
            for (num, word) in enumerate(words)
                if startswith(line[c:end], word)
                    line = string(line[begin:c-1], num, line[c:end])
                    c += 1
                end
            end
            c += 1
        end
        return line
    end

    return @pipe lines |> map(fixup, _) |> solve1
end

check("Day $day.2 Sample",  () -> solve2(sample2), 281)
check("Day $day.2 Problem", () -> solve2(input), 53868)
